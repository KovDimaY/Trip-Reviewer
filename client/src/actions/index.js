import axios from 'axios';

import {
    GET_TRIPS,
    GET_TRIP_W_REVIEWER,
    CLEAR_TRIP_W_REVIEWER,
    USER_LOGIN,
    USER_AUTH,
    ADD_TRIP,
    CLEAR_NEW_TRIP,
    GET_USER_REVIEWS
} from '../constants/action-names';

export function getTrips(
    limit = 10,
    start = 0,
    order = 'asc',
    list = []
) {
    const request = axios
        .get(`/api/getManyTrips?limit=${limit}&skip=${start}&order=${order}`)
        .then(response => [...list, ...response.data]);

    return {
        type: GET_TRIPS,
        payload: request
    };
}

export function getTripWithReviewer(id) {
    const request = axios.get(`/api/getTrip?id=${id}`);

    return (dispatch) => {
        request.then(({ data }) => {
            const trip = data;

            axios
                .get(`/api/getReviewer?id=${trip.ownerId}`)
                .then(({ data }) => {
                    const response = {
                        trip,
                        reviewer: data
                    };

                    dispatch({
                        type: GET_TRIP_W_REVIEWER,
                        payload: response
                    });
                });
        });
    };
}

export function getUserReviews(userId) {
    const request = axios
        .get(`/api/getUserReviews?user=${userId}`)
        .then(response => response.data);

    return {
        type: GET_USER_REVIEWS,
        payload: request
    };
}

export function clearTripWithReviewer() {
    return {
        type: CLEAR_TRIP_W_REVIEWER,
        payload: {
            trip: null,
            reviewer: null
        }
    };
}

export function loginUser({ email, password }) {
    const request = axios
        .post('/api/login',{ email, password })
        .then(response => response.data);

    return {
        type: USER_LOGIN,
        payload: request
    };
}

export function auth() {
    const request = axios
        .get('/api/auth')
        .then(response => response.data);

    return {
        type: USER_AUTH,
        payload: request
    }
}

export function addTrip(trip) {
    const request = axios
        .post('/api/trip', trip)
        .then(response => response.data);

    return {
        type: ADD_TRIP,
        payload: request
    };
}

export function clearNewTrip() {
    return {
        type: CLEAR_NEW_TRIP,
        payload: null
    };
}