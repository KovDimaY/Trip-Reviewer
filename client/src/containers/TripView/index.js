import React, { Component } from 'react';
import { getTripWithReviewer, clearTripWithReviewer } from '../../actions';
import { connect } from 'react-redux';

class TripView extends Component {
    componentWillMount(){
        this.props.dispatch(getTripWithReviewer(this.props.match.params.id))
    }

    componentWillUnmount(){
        this.props.dispatch(clearTripWithReviewer())
    }

    renderTrip = (trips) => (
        trips.current 
            ? <div className="br_container">
                    <div className="br_header">
                        <h2>{trips.current.title}</h2>
                        <h5>{trips.current.author}</h5>
                        <div className="br_reviewer">
                            <span>Review by:</span> {trips.reviewer.name} {trips.reviewer.lastname}
                        </div>
                    </div>
                    <div className="br_review">
                        {trips.current.review}
                    </div>
                    <div className="br_box">
                        <div className="left">
                            <div>
                                <span>Duration:</span> {trips.current.duration} days
                            </div>
                            <div>
                                <span>Price:</span> {trips.current.price}
                            </div>
                        </div>
                        <div className="right">
                            <span>Rating</span>
                            <div>{trips.current.rating}/5</div>
                        </div>
                    </div>
                </div>
            :   null
    )

    render() {
        const { trips } = this.props;
        return (
            <div>
                {this.renderTrip(trips)}
            </div>
        );
    }
}

function mapStateToProps(state){
    return {
        trips: state.trips
    };
};

export default connect(mapStateToProps)(TripView);