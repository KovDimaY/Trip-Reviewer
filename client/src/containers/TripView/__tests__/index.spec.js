import React from 'react';
import configureStore from 'redux-mock-store';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';

import TripView from '..';
import { clearTripWithReviewer } from '../../../actions';

jest.mock('./../../../actions', () => ({
  getTripWithReviewer: jest.fn(() => ({
    type: 'getTripWithReviewer',
  })),
  clearTripWithReviewer: jest.fn(() => ({
    type: 'clearTripWithReviewer',
  })),
}));

const mockStore = configureStore();

const mockComponent = (initialState = {}, props) => {
  const store = mockStore(initialState);

  return (
    <TripView {...props} store={store} />
  );
};

describe('<TripView />', () => {
  it('should render component with empty trips', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };

    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with current trip', () => {
    const initialState = {
      trips: {
        current: {
          title: 'title',
          author: 'author',
          review: 'review',
          duration: 'duration',
          price: 'price',
          rating: 'rating',
        },
        reviewer: {
          name: 'name',
          lastname: 'lastname',
        },
      },
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should dispatch clearTripWithReviewer when componentWillUnmount is called', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillUnmount();

    expect(clearTripWithReviewer).toHaveBeenCalled();
  });
});
