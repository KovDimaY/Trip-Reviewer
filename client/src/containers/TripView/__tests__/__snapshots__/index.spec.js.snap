// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripView /> should render component with current trip of 1 day 1`] = `
<div>
  <div
    className="trip-view-container"
  >
    <div
      className="trip-view-header"
    >
      <h2>
        title
      </h2>
      <h5>
        country
      </h5>
      <div
        className="reviewer"
      >
        <span>
          Author:
        </span>
         name lastname
      </div>
    </div>
    <div
      className="trip-view-description"
    >
      <Editor
        editorClassName="editor-self"
        editorState={Object {}}
        readOnly={true}
        toolbarClassName="editor-toolbar"
        toolbarHidden={true}
        wrapperClassName="editor-wrapper"
      />
    </div>
    <div
      className="trip-view-box"
    >
      <div
        className="left"
      >
        <div>
          <span>
            Duration:
          </span>
           1 day
        </div>
        <div>
          <span>
            Expences:
          </span>
           $expences 
        </div>
      </div>
      <div
        className="right"
      >
        <span>
          Rating
        </span>
        <div>
          rating
          /5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TripView /> should render component with current trip of 10 day 1`] = `
<div>
  <div
    className="trip-view-container"
  >
    <div
      className="trip-view-header"
    >
      <h2>
        title
      </h2>
      <h5>
        country
      </h5>
      <div
        className="reviewer"
      >
        <span>
          Author:
        </span>
         name lastname
      </div>
    </div>
    <div
      className="trip-view-description"
    >
      <Editor
        editorClassName="editor-self"
        editorState={Object {}}
        readOnly={true}
        toolbarClassName="editor-toolbar"
        toolbarHidden={true}
        wrapperClassName="editor-wrapper"
      />
    </div>
    <div
      className="trip-view-box"
    >
      <div
        className="left"
      >
        <div>
          <span>
            Duration:
          </span>
           10 days
        </div>
        <div>
          <span>
            Expences:
          </span>
           $expences 
        </div>
      </div>
      <div
        className="right"
      >
        <span>
          Rating
        </span>
        <div>
          rating
          /5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TripView /> should render component with empty trips 1`] = `<div />`;
